src/message.rs

```diff
@@ -16,16 +20,40 @@ pub struct Message {
     pub location: Location,
     pub end_location: Location,
     pub filename: String,
+    pub source: Option<String>,
+    pub range: Option<(usize, usize)>,
+    pub show_source: bool,
 }
 
 impl Message {
-    pub fn from_check(filename: String, check: Check) -> Self {
+    pub fn from_check(
+        filename: String,
+        check: Check,
+        locator: &SourceCodeLocator,
+        show_source: bool,
+    ) -> Self {
+        let source = locator.slice_source_code_range(&Range {
+            location: Location::new(check.location.row(), 0),
+            end_location: Location::new(check.end_location.row() + 1, 0),
+        });
+        let num_chars_in_range = locator
+            .slice_source_code_range(&Range {
+                location: check.location,
+                end_location: check.end_location,
+            })
+            .len();
```

Comment:
I think this needs to be `.chars().len()`, or we need to use the `Rope` in `locator` to get the start and end _character_ positions.